File: src/main/java/org/academy/api/common/profiler/ProfilerSampler.java
package org.academy.api.common.profiler;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.lang.management.ManagementFactory;
import java.lang.management.ThreadInfo;
import java.lang.management.ThreadMXBean;
import java.util.*;
import java.util.concurrent.ConcurrentHashMap;

public final class ProfilerSampler {
    private static final Logger LOGGER = LoggerFactory.getLogger("AcademyProfiler");
    private static final ThreadMXBean threadBean = ManagementFactory.getThreadMXBean();
    private static final ConcurrentHashMap<Long, ThreadRef> threads = new ConcurrentHashMap<>();
    private static final ConcurrentHashMap<Long, SampledCallTree> trees = new ConcurrentHashMap<>();
    private static final int MAX_DEPTH = 128;

    private static volatile long intervalMicros = 1000L;
    private static volatile boolean running = false;
    private static volatile boolean paused = false;
    private static volatile long captureStartNanos = System.nanoTime();
    private static volatile boolean everStarted = false;

    private static volatile Thread worker = null;
    private static final Object lock = new Object();

    private ProfilerSampler() {
    }

    public static boolean isRunning() {
        return running;
    }

    public static boolean isPaused() {
        return paused;
    }

    public static boolean hasData() {
        return everStarted;
    }

    public static long getIntervalMicros() {
        return intervalMicros;
    }

    public static List<ThreadRef> threadRefs() {
        List<ThreadRef> refs = new ArrayList<>(threads.values());
        refs.sort(Comparator.comparing(ThreadRef::getName));
        return refs;
    }

    public static ThreadRef registerThread(Thread thread) {
        long threadId = thread.threadId();
        ThreadRef ref = threads.computeIfAbsent(threadId, id -> new ThreadRef(id, thread.getName()));
        trees.computeIfAbsent(threadId, id -> new SampledCallTree(id, thread.getName()));
        return ref;
    }

    public static void unregisterThread(long threadId) {
        threads.remove(threadId);
        trees.remove(threadId);
    }

    public static void setThreadEnabled(long threadId, boolean enabled) {
        ThreadRef ref = threads.get(threadId);
        if (ref != null) {
            ref.enabled = enabled;
        }
    }

    public static void start(long intervalMicros) {
        synchronized (lock) {
            if (running) {
                return;
            }
            running = true;
            paused = false;
            everStarted = true;
            ProfilerSampler.intervalMicros = Math.max(100L, Math.min(1_000_000L, intervalMicros));
            captureStartNanos = System.nanoTime();
            resetInternal();
            Thread thread = new Thread(ProfilerSampler::loop, "Academy Profiler Sampler");
            thread.setDaemon(true);
            worker = thread;
            thread.start();
        }
    }

    public static void stop() {
        synchronized (lock) {
            if (!running) {
                return;
            }
            running = false;
            if (worker != null) {
                worker.interrupt();
            }
            worker = null;
        }
    }

    public static void pause() {
        paused = true;
    }

    public static void resume() {
        paused = false;
    }

    public static void reset() {
        synchronized (lock) {
            resetInternal();
        }
    }

    private static void resetInternal() {
        captureStartNanos = System.nanoTime();
        for (SampledCallTree tree : trees.values()) {
            tree.reset();
        }
    }

    public static double elapsedSeconds() {
        return (System.nanoTime() - captureStartNanos) / 1e9;
    }

    private static void loop() {
        while (running) {
            if (!paused) {
                try {
                    sampleOnce();
                } catch (Throwable t) {
                    LOGGER.warn("Sampler iteration failed", t);
                }
            }
            long us = intervalMicros;
            long sleepMs = us / 1000;
            int sleepNs = (int) ((us % 1000) * 1000);
            try {
                Thread.sleep(sleepMs, sleepNs);
            } catch (InterruptedException e) {
                break;
            }
        }
    }

    private static void sampleOnce() {
        List<Long> enabledIds = new ArrayList<>();
        for (ThreadRef ref : threads.values()) {
            if (ref.enabled) {
                enabledIds.add(ref.id);
            }
        }
        if (enabledIds.isEmpty()) {
            return;
        }
        long[] ids = enabledIds.stream().mapToLong(Long::longValue).toArray();
        ThreadInfo[] infos;
        try {
            infos = threadBean.getThreadInfo(ids, MAX_DEPTH);
        } catch (Throwable t) {
            return;
        }
        for (int i = 0; i < ids.length; i++) {
            ThreadInfo info = infos[i];
            if (info == null) {
                continue;
            }
            StackTraceElement[] stack = info.getStackTrace();
            if (stack.length == 0) {
                continue;
            }
            SampledCallTree tree = trees.get(ids[i]);
            if (tree != null) {
                tree.insert(stack);
            }
        }
    }

    public static SamplerSnapshot snapshot() {
        Map<Long, SampledThreadView> perThread = new LinkedHashMap<>();
        for (Map.Entry<Long, SampledCallTree> entry : trees.entrySet()) {
            SampledCallTree tree = entry.getValue();
            long total = tree.totalSamples();
            perThread.put(entry.getKey(), new SampledThreadView(
                    entry.getKey(),
                    tree.getThreadName(),
                    total,
                    cloneNode(tree.getRoot(), total)
            ));
        }
        long totalSamples = 0;
        for (SampledThreadView view : perThread.values()) {
            totalSamples += view.samples();
        }
        return new SamplerSnapshot(perThread, totalSamples, elapsedSeconds());
    }

    private static SampledNode cloneNode(SampledCallNode node, long totalSamples) {
        List<SampledCallNode> rawChildren = new ArrayList<>(node.getChildren().values());
        rawChildren.sort((a, b) -> Long.compare(b.getSamples().sum(), a.getSamples().sum()));
        List<SampledNode> children = new ArrayList<>();
        for (SampledCallNode child : rawChildren) {
            children.add(cloneNode(child, totalSamples));
        }
        return new SampledNode(
                node.getLabel(),
                node.getSamples().sum(),
                node.getSelfSamples().sum(),
                children,
                totalSamples
        );
    }

    public static class ThreadRef {
        private final long id;
        private final String name;
        public volatile boolean enabled = true;

        public ThreadRef(long id, String name) {
            this.id = id;
            this.name = name;
        }

        public long getId() {
            return id;
        }

        public String getName() {
            return name;
        }

        public boolean isEnabled() {
            return enabled;
        }

        public void setEnabled(boolean enabled) {
            this.enabled = enabled;
        }
    }
}

File: src/main/kotlin/org/academy/api/common/profiler/ProfilerSampler.kt
package org.academy.api.common.profiler

import org.slf4j.LoggerFactory
import java.lang.management.ManagementFactory
import java.util.concurrent.ConcurrentHashMap

/**
 * 基于 ThreadMXBean 的采样式性能剖析器（Spark 风格）。
 *
 * 在独立守护线程上周期性地抓取已注册线程的栈，聚合进 [SampledCallTree]。
 * 完全自包含，不依赖 MC 的 Profiler / Tracy / 外部环境。
 */
object ProfilerSampler {
    private val LOGGER = LoggerFactory.getLogger("AcademyProfiler")
    private val threadBean = ManagementFactory.getThreadMXBean()
    private val threads = ConcurrentHashMap<Long, ThreadRef>()
    private val trees = ConcurrentHashMap<Long, SampledCallTree>()

    private const val MAX_DEPTH = 128

    @Volatile
    var intervalMicros: Long = 1000L
        private set

    @Volatile
    private var running = false

    @Volatile
    private var paused = false

    @Volatile
    private var captureStartNanos = System.nanoTime()

    @Volatile
    private var everStarted = false

    private var worker: Thread? = null
    private val lock = Any()

    class ThreadRef(val id: Long, val name: String) {
        @Volatile
        var enabled: Boolean = true
    }

    val isRunning: Boolean get() = running
    val isPaused: Boolean get() = paused

    /** 是否启动过（保留停止后的数据，供 UI / 导出查看）。 */
    val hasData: Boolean get() = everStarted

    fun threadRefs(): List<ThreadRef> = threads.values.sortedBy { it.name }

    fun registerThread(thread: Thread): ThreadRef {
        val ref = threads.computeIfAbsent(thread.id) { ThreadRef(thread.id, thread.name) }
        trees.computeIfAbsent(thread.id) { SampledCallTree(thread.id, thread.name) }
        return ref
    }

    fun unregisterThread(threadId: Long) {
        threads.remove(threadId)
        trees.remove(threadId)
    }

    fun setThreadEnabled(threadId: Long, enabled: Boolean) {
        threads[threadId]?.enabled = enabled
    }

    fun start(intervalMicros: Long) {
        synchronized(lock) {
            if (running) return
            running = true
            paused = false
            everStarted = true
            this.intervalMicros = intervalMicros.coerceIn(100L, 1_000_000L)
            captureStartNanos = System.nanoTime()
            resetInternal()
            worker = Thread({ loop() }, "Academy Profiler Sampler").apply {
                isDaemon = true
                start()
            }
        }
    }

    fun stop() {
        synchronized(lock) {
            if (!running) return
            running = false
            worker?.interrupt()
            worker = null
        }
    }

    fun pause() {
        paused = true
    }

    fun resume() {
        paused = false
    }

    fun reset() {
        synchronized(lock) { resetInternal() }
    }

    private fun resetInternal() {
        captureStartNanos = System.nanoTime()
        trees.values.forEach { it.reset() }
    }

    fun elapsedSeconds(): Double = (System.nanoTime() - captureStartNanos) / 1e9

    private fun loop() {
        while (running) {
            if (!paused) {
                try {
                    sampleOnce()
                } catch (t: Throwable) {
                    LOGGER.warn("Sampler iteration failed", t)
                }
            }
            val us = intervalMicros
            val sleepMs = us / 1000
            val sleepNs = (us % 1000) * 1000
            try {
                Thread.sleep(sleepMs, sleepNs.toInt())
            } catch (e: InterruptedException) {
                break
            }
        }
    }

    private fun sampleOnce() {
        val enabledIds = threads.values.filter { it.enabled }.map { it.id }.toLongArray()
        if (enabledIds.isEmpty()) return
        val infos = try {
            threadBean.getThreadInfo(enabledIds, MAX_DEPTH)
        } catch (t: Throwable) {
            null
        } ?: return
        for (i in enabledIds.indices) {
            val info = infos[i] ?: continue
            val stack = info.stackTrace
            if (stack.isEmpty()) continue
            trees[enabledIds[i]]?.insert(stack)
        }
    }

    fun snapshot(): SamplerSnapshot {
        val perThread = trees.mapValues { (id, tree) ->
            val total = tree.totalSamples()
            SampledThreadView(
                id = id,
                name = tree.threadName,
                samples = total,
                root = cloneNode(tree.root, total),
            )
        }
        return SamplerSnapshot(
            threads = perThread,
            totalSamples = perThread.values.sumOf { it.samples },
            durationSeconds = elapsedSeconds(),
        )
    }

    private fun cloneNode(node: SampledCallNode, totalSamples: Long): SampledNode {
        val children = node.children.values
            .sortedByDescending { it.samples.sum() }
            .map { cloneNode(it, totalSamples) }
        return SampledNode(
            label = node.label,
            samples = node.samples.sum(),
            selfSamples = node.selfSamples.sum(),
            children = children,
            totalSamples = totalSamples,
        )
    }
}

class SampledNode(
    val label: String,
    val samples: Long,
    val selfSamples: Long,
    val children: List<SampledNode>,
    val totalSamples: Long,
) {
    val samplesPercent: Double get() = if (totalSamples > 0) samples * 100.0 / totalSamples else 0.0
    val selfPercent: Double get() = if (totalSamples > 0) selfSamples * 100.0 / totalSamples else 0.0
}

class SampledThreadView(
    val id: Long,
    val name: String,
    val samples: Long,
    val root: SampledNode,
)

class SamplerSnapshot(
    val threads: Map<Long, SampledThreadView>,
    val totalSamples: Long,
    val durationSeconds: Double,
)
